docs/proposals/20200511-clusterctl-extensible-template-processing.md

```diff
@@ -0,0 +1,344 @@
+  cluster, URL, Github Repository, Local Repository and even the overrides
+  directory. To ensure compatibility, we’ll need to establish a compression
+  format like tar, gzip or zip.
+
+## Alternatives
+- An alternative to allowing clusterctl handle templates would be to just use
+  another tool upstream to generate the various templates with variables
+  defined. Then use clusterctl to consume that template and use the default
+  variable substitution templating solution.
+
+
+## Upgrade Strategy
+
+
+Any individual template engine libraries would be upgraded via go modules
+preferably using semver standards.
+
+
+## Additional Details
+
+
+Currently, cluster-api-provider-azure is using kustomize to author their
+cluster template flavors. See [issue 540] for more details. Maybe we can
+collect input from them to understand their strategy if they need to
+manage their template variations.
+
+
+## Implementation History
+
+
+- [wfernandes] 04/22/2020: Proposed idea in an issue or [community meeting]
+- [wfernandes/fabrizio.pandini] 04/27/2020: Compile a [CAEP Google Doc] following the CAEP template
+- [wfernandes] 05/06/2020: First round of feedback from community
+- [wfernandes ] 05/13/2020: Present proposal at a [community meeting]
+- [wfernandes ] 05/11/2020: Open proposal PR
+
+
+## Open Questions
+<!-- Add your questions here -->
+
+1. [wfernandes] How do we pass in values to the YamlProcessor? Currently,
+   clusterctl retrieves values from environment variables and the config; some
+   values from flags. However, for a template engine like ytt, it can specify
+   nested values in the `values.yaml`.
+   
+   For example,
+   
+   ```
+   aws:
+    secret_key: foo
+    access_key: bar
+   ```
+   
+   And ytt can reference these values as flags in the form `aws.secret_key`.
+   Should we provide other means of variable input to clusterctl?
+   
+   Notes:
+   Ytt is able to read environment variables. However, it does have some caveats.
+   [More information can be found here][ytt-data-values].
+   
+2. [wfernandes] We need to define a format or convention for retrieving multiple template files.
+    Currently, clusterctl supports template retrieval from a ConfigMap in a
+    cluster, URL, Github release artifacts and local file system. We default to
+    retrieving provider cluster templates from github release assets. In order to
+    avoid running into github quota limits on retrieving asset files we can pull a
+    compressed file with all the template contents.
+    
+      As mentioned above in the *Constraints* and *Implementation* section a suggestion
+      for convention would be
+      `cluster-template-<template-engine-type>.<compression-format>`. Compression
+      format could be tar, gzip, zip, etc.
+    
+3. [wfernandes] Do we want to extend the usage of other yaml processors to
+   include infrastructure components? That is, do we want to stick with using
```

Comment:
One concrete thing I'd like to see is a way to make certain portions of provider templates optional. For example, when bootstrapping a new management cluster, you typically need to use a `Secret` to contain the credentials you want to use when communicating with your infrastructure provider (vSphere, AWS, Azure, etc.). After pivoting the Cluster API infrastructure deployment to the real new management cluster, in many situations, you'll want to use credentials granted to the underlying infrastructure machine (e.g. an AWS IAM instance profile) instead of a static credentials `Secret`. We don't necessarily have to figure this out as part of this initial work, but it is something I'd like to get to eventually.

Reply:
+1
Another possible use case is the management of controller flags
(see https://github.com/kubernetes-sigs/cluster-api/issues/3045)

Reply:
Currently, the configuration required by the infrastructure components are directly related to credential management. For example, `AWS_B64_ENCODED_CREDENTIALS`, `AZURE_CLIENT_SECRET_B64`, `VSPHERE_PASSWORD`. Since there seems to be another effort to address the sharing of secrets in the future and possibly along with the management of controller flags, I suggest as part of this proposal, we use the default yaml processor for the infrastructure components.
That is, `clusterctl init` will operate as usual with variable substitution. It will not take the flag `--experimental-template-engine`.